### DISCUSSION TITLE
Providing custom starting source position to parser.

### DISCUSSION THREAD

**Discussion by forbisc** [tree-sitter/tree-sitter#1175] [Q&A] [answered]
It seems possible to provide a custom starting position to a piece of source code by using `set_included_ranges` and `parse_with` to convince tree sitter that the given source is located at a particular position.  However, it doesn't work if the range starts at byte 0, because this line (which modifies extent) is only hit when the included range is beyond the current position.  

https://github.com/tree-sitter/tree-sitter/blob/12207c1f70dd9e5d4bfb168cbb4f942f8a4f181b/lib/src/lexer.c#L116

I can workaround the issue by adding a dummy byte at the start and then using setting the included range to start at byte 1.  Then it will set 'extent' to the value specified in the included range.  

My question is: Am I misusing set_included_ranges or is this a bug?  Should the lexer update extent if the current position is at the very beginning of the included range?

Basically I think that this line should read 
      if (included_range->start_byte >= position.bytes) {

https://github.com/tree-sitter/tree-sitter/blob/12207c1f70dd9e5d4bfb168cbb4f942f8a4f181b/lib/src/lexer.c#L113

**Comment by ahlinc** [comment, marked as answer]
I think you found the bug in the core lib, a PR is welcome but such fixes to the core lib usually require a clear test that fails on the bug. All tests are in Rust and the test needs to be added somewhere around of: https://github.com/tree-sitter/tree-sitter/blob/12207c1f70dd9e5d4bfb168cbb4f942f8a4f181b/cli/src/tests/parser_test.rs#L784

**Reply by forbisc** [reply]
I created #1177 .
